[CK_BUILDER] Test and fix instance traits utils.

shumway opened the pull request:
The PR adds much needed unit testing for instance traits utils.

The `tuple_name` function was completely broken, it returned "EmptyTuple" for any input. Implemented that and added a `sequence_name` function, too, though that may not be needed if we always convert sequences to std::vector. 

Added names to ops that needed them to pass unit tests.

Also did some cleanup with `using enum`.

shumway committed:
Refactor instance_traits_util and add unit tests tests

bartekxk commented on experimental/builder/test/test_instance_traits_util.cpp:
```diff
@@ -0,0 +1,260 @@
+                                               layout_name<conv::GNKHW>(),
+                                               layout_name<conv::GNHWK>(),
+                                               // Strided layouts
+                                               layout_name<conv::G_NHW_C>(),
```
I think that we dont use such layouts. Strides are modified by passing other vales in strides parameter.

shumway replied:
Marked as deprecated. Keeping until we've removed the instances.

bartekxk commented on experimental/builder/test/test_instance_traits_util.cpp:
```diff
@@ -0,0 +1,260 @@
+TEST(InstanceTraitsUtil, LayoutNameReturnsCorrectStringsForGemmLayouts)
+{
+    namespace gemm = ck::tensor_layout::gemm;
+    EXPECT_THAT((std::vector<std::string_view>{layout_name<gemm::RowMajor>(),
```
For old ck it doesnt has an impact but in the old ck it has. For backward weight layout is Col -Row and for bwd data Row Row

shumway replied:
Resolved in chat.

bartekxk commented on experimental/builder/test/test_instance_traits_util.cpp:
```diff
@@ -0,0 +1,260 @@
+{
+    namespace conv = ck::tensor_layout::convolution;
+    EXPECT_THAT((std::vector<std::string_view>{// Input tensor layouts (packed)
+                                               layout_name<conv::NCHW>(),
```
Can we drop non grouped layouts?

shumway replied:
We still need these for reflection on the current code, but marked them deprecated.

aosewski commented on experimental/builder/include/ck_tile/builder/reflect/instance_traits_util.hpp:
```diff
@@ -55,20 +48,36 @@ consteval std::string_view type_name()
+
+// Concept that checks valid layout types
+template <typename T>
+concept IsLayoutType = std::is_base_of_v<ck::tensor_layout::BaseTensorLayout, T> && requires {
```
You should check here as well for old CK BaseTensorLayout struct.

shumway replied:
Done.

aosewski commented on experimental/builder/include/ck_tile/builder/reflect/instance_traits_util.hpp:
```diff
@@ -22,19 +22,12 @@
+// Implementation detail for type name mapping
+// This is the single source of truth for supported data types that
+// returns an empty string to indicate an unsupported type.
+namespace detail {
```
detail::detail ? You're already under `ck_tile::reflect::detail` namespace... maybe then anonymous would be better ?

shumway replied:
Moved to detail::impl. It's not great, but I'm not sure how else to hide this.

shumway committed:
Address reviewer comments.

Just adds some TODOs to indicate deprecated layouts in our reflection. Our strategy is to leave the reflection code broad (covering deprecated features), but keep the builder concepts narrow. Once we've removed deprecated features from all instances, we can remove them from reflection.

Also add a comment to the cmake to explain the unit test target test_conv_builder.

shumway committed:
Merge branch 'develop' into jshumway/instance-trait-tuples

shumway committed:
Merge branch 'develop' into jshumway/instance-trait-tuples

shumway committed:
Addressed more reviewer comments.

shumway committed:
Remove duplicate PassThrough::name

Accidentally added this field to the end of the struct, too. The `name` field should be a the start of the struct for consistency.
